client/components/Editor/plugins/inputRules.ts

```diff
@@ -104,31 +104,50 @@ const inlineMathRule = (
 // textblock starting with two dollar signs into a math block.
 const blockMathRule = (nodeType) => makeBlockMathInputRule(REGEX_BLOCK_MATH_DOLLARS, nodeType);
 
-const HTTP_MAILTO_REGEX = new RegExp(
-	// eslint-disable-next-line no-useless-escape
-	/(?:(?:(https|http|ftp)+):\/\/)?(?:\S+(?::\S*)?(@))?(?:(?:([a-z0-9][a-z0-9\-]*)?[a-z0-9]+)(?:\.(?:[a-z0-9\-])*[a-z0-9]+)*(?:\.(?:[a-z]{2,})(:\d{1,5})?))(?:\/[^\s]*)?\s$/,
-);
+const emailOrUriRegexBase =
+	'(?<emailOrUri>(?:(?:(https|http|ftp)+)://)?(?:\\S+(?::\\S*)?(?<atSign>@))?(?:(?:([a-z0-9][a-z0-9-]*)?[a-z0-9]+)(?:\\.(?:[a-z0-9-])*[a-z0-9]+)*(?:\\.(?:[a-z]{2,})(:\\d{1,5})?))(?:/[^\\s]*)?)';
 
+// Export a version of the regex and handler so that we can reuse this logic in a custom command
+// mapped to the `enter` key, because input rules don't work across nodes
+export const EMAIL_OR_URI_REGEX = new RegExp(`${emailOrUriRegexBase}$`);
+
+const EMAIL_OR_URI_REGEX_WITH_SPACE = new RegExp(`${emailOrUriRegexBase}(?<whitespace>\\s)$`);
+
+// Determine if the matched content is a url or an email and add a link mark to it
+export const linkRuleHandler = (
+	markType: MarkType,
+	appendWhitespace = false,
+	transaction?: Transaction,
+) => {
+	return (state: EditorState, match: RegExpMatchArray, start: number, end: number) => {
+		const resolvedStart = state.doc.resolve(start);
+		const tr = transaction ?? state.tr;
+		if (!resolvedStart.parent.type.allowsMarkType(markType)) {
+			return tr;
+		}
+		const emailOrUri = match.groups!.emailOrUri;
+
+		const href = `${match.groups!.atSign ? 'mailto:' : ''}${emailOrUri}`;
```

Comment:
Adding a mailto link inserts mailto in the link twice
[Screencast from 05-01-2023 05:27:38 PM.webm](https://user-images.githubusercontent.com/34730449/235534695-ae898d45-a0f8-4dc6-8c71-fd84a4e7cd60.webm)

Reply:
Since this is also not a regression and I'm trying to close stuff out this sprint, I'd like to just write up issues for both of these changes instead of putting them in this PR.

In this case, I'm also not really sure what the right behavior is. Should we be linkifying the entire string of `mailto:user@example.com`?